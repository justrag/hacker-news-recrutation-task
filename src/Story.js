import React from 'react';
import { connect } from 'react-redux';
import {
  Badge,
  Icon,
  IconButton,
  Card,
  CardText,
  CardTitle,
  CardActions,
  Button,
  CardMenu
} from 'react-mdl';
import { chooseStory, showModal } from './actions';
import { getChosenStory } from './selectors';

const localDateTime = unixtime => {
  const newDate = new Date(1000 * unixtime);
  /*
  return newDate.toLocaleDateString(undefined, {
    day: 'numeric',
    month: 'long',
    year: 'numeric',
    hour: 'numeric',
    minute: 'numeric'
  });
  */
  return newDate.toDateString() + ' ' + newDate.toTimeString();
};

const domainFromUrl = url => {
  const link = document.createElement('a');
  link.setAttribute('href', url);
  return link.hostname;
};

/*
const Story = ({
  id,
  title,
  by,
  score,
  time,
  descendants,
  url,
  chooseStory,
  showModal,
  chosen
}) =>
  id === chosen ? (
    <ListItem twoLine>
      <ListItemContent
        subtitle={`${score} points by ${by} at ${localDateTime(time)}; ${
          descendants
        } comments:`}
      >
        <span onClick={() => showModal(url)}>
          {title} ({url})
        </span>
      </ListItemContent>
    </ListItem>
  ) : (
    <ListItem>
      <ListItemContent avatar="web">{title}</ListItemContent>
      <ListItemAction>
        <Icon name="arrow forward" onClick={() => chooseStory(id)} />
      </ListItemAction>
    </ListItem>
  );
  */

const cardStyle = {
  width: '95%',
  margin: 'auto',
  marginBottom: '10px',
  minHeight: '1px'
};
const cardMenuStyle = { color: '#795548' };
const Story = ({
  id,
  title,
  by,
  score,
  time,
  descendants,
  url,
  chooseStory,
  showModal,
  chosen
}) =>
  id === chosen ? (
    <Card shadow={3} style={cardStyle}>
      <CardTitle>{title}</CardTitle>
      <CardText>
        <Badge text={score} overlap>
          <Icon name="favorite" />
        </Badge>
        <Badge text={descendants} overlap>
          <Icon name="comment" />
        </Badge>
        <Icon name="person" /> {by} <Icon name="date_range" />{' '}
        {localDateTime(time)}
      </CardText>
      <CardActions border>
        <Button onClick={() => showModal(url)} colored>
          {domainFromUrl(url)}
        </Button>
      </CardActions>
    </Card>
  ) : (
    <Card shadow={1} style={cardStyle}>
      <CardText>{title}</CardText>
      <CardMenu style={cardMenuStyle}>
        <IconButton name="arrow_back" onClick={() => chooseStory(id)} />
      </CardMenu>
    </Card>
  );

const mapStateToProps = state => ({
  chosen: getChosenStory(state)
});

export default connect(mapStateToProps, { chooseStory, showModal })(Story);
