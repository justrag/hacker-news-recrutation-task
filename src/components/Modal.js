import React from 'react';
import { connect } from 'react-redux';
import { IconButton, Content } from 'react-mdl';
import { getModalHref } from '../selectors/';
import { hideModal } from '../actions/';

const modalStyle = {
  position: 'fixed',
  backgroundColor: '#795548',
  zIndex: 1,
  left: '2%',
  top: '10%',
  width: '95%',
  height: '90%',
  //  overflow: 'auto',
  border: 'thin solid darkgrey'
};
const modalContentStyle = {
  backgroundColor: '#fefefe',
  margin: '15px auto',
  padding: '10px',
  border: '1px solid #888',
  width: '90%',
  height: '90%',
  display: 'block'
};

const closeButtonStyle = {
  float: 'right'
};

export const Modal = ({ href, hideModal }) =>
  href && (
    <Content style={modalStyle}>
      <IconButton
        style={closeButtonStyle}
        onClick={hideModal}
        name="close"
        colored
      />
      <iframe style={modalContentStyle} src={href} />
    </Content>
  );

const mapStateToProps = state => ({ href: getModalHref(state) });

export default connect(mapStateToProps, { hideModal })(Modal);
