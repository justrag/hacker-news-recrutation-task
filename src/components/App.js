import React from 'react';
import { Layout, Header, Footer } from 'react-mdl';
import Modal from './Modal';
import Stories from './Stories';

const layoutStyle = {
  background: 'url(Atari-800xl.jpg) center / cover'
};
//pic from http://bestandworstever.blogspot.com/2012/12/best-atari-system-ever-800xl.html

const titleStyle = { color: '#ddd' };
export const App = () => (
  <div>
    <Layout fixedHeader style={layoutStyle}>
      <Header title={<span style={titleStyle}>HackerNews</span>} />
      <Stories />
      <Modal />
      <Footer size="mini">
        <span>made by rag in 2018</span>
      </Footer>
    </Layout>
  </div>
);

export default App;
