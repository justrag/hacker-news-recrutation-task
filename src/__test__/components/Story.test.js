import React from 'react';
import { render } from 'enzyme';
import { Story } from '../../components/Story';

const chosenStory = {
  by: 'tgragnato',
  descendants: 3,
  id: 11,
  score: 12,
  time: 1515459435,
  title: 'Economics is overdue for a quantum revolution',
  url:
    'https://aeon.co/essays/has-the-time-come-for-a-quantum-revolution-in-economics',
  chooseStory: jest.fn(),
  showModal: jest.fn(),
  chosen: 11
};
const notChosenStory = { ...chosenStory, id: 12 };

describe('Component: Story', () => {
  it('renders the chosen story correctly', () => {
    const test = render(<Story {...chosenStory} />);
    expect(test).toMatchSnapshot();
  });

  it('renders the regular story correctly', () => {
    const test = render(<Story {...notChosenStory} />);
    expect(test).toMatchSnapshot();
  });
});
