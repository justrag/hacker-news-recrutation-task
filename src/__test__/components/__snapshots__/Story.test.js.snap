// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Story renders the chosen story correctly 1`] = `
<div
  class="mdl-card mdl-shadow--6dp"
  style="width:95%;margin:auto;margin-bottom:10px;min-height:1px"
>
  <div
    class="mdl-card__title"
  >
    <h2
      class="mdl-card__title-text"
    >
      Economics is overdue for a quantum revolution
    </h2>
  </div>
  <div
    class="mdl-card__supporting-text"
  >
    <i
      class="material-icons mdl-badge mdl-badge--overlap"
      data-badge="12"
    >
      favorite
    </i>
    <i
      class="material-icons mdl-badge mdl-badge--overlap"
      data-badge="3"
    >
      comment
    </i>
    <i
      class="material-icons"
    >
      person
    </i>
     tgragnato 
    <i
      class="material-icons"
    >
      date_range
    </i>
     Tue Jan 09 2018 01:57:15 GMT+0100 (CET)
  </div>
  <div
    class="mdl-card__actions mdl-card--border"
  >
    <button
      class="mdl-button mdl-js-button mdl-button--colored"
    >
      aeon.co
    </button>
  </div>
</div>
`;

exports[`Component: Story renders the regular story correctly 1`] = `
<div
  class="mdl-card mdl-shadow--3dp"
  style="width:95%;margin:auto;margin-bottom:10px;min-height:1px"
>
  <div
    class="mdl-card__supporting-text"
  >
    Economics is overdue for a quantum revolution
  </div>
  <div
    class="mdl-card__menu"
    style="color:#795548"
  >
    <button
      class="mdl-button mdl-js-button mdl-button--icon"
    >
      <i
        class="material-icons"
      >
        arrow_back
      </i>
    </button>
  </div>
</div>
`;
